### ISSUE TITLE
BufferUnderflowException when using HLL in TopN query

### ISSUE THREAD

**Issue by kaijianding** [druid-io/druid#4199]
I found similar issue reported at https://github.com/druid-io/druid/issues/3560  it's marked as fixed in 0.9.2
But I still met this issue in 0.9.2, I wonder is it still happening in 0.10.0?
Here is the trace from historical log:

> , exception=java.nio.BufferUnderflowException, query=TopNQuery{dataSource='dws_wl_hash_online', dimensionSpec=DefaultDimensionSpec{dimension='query', outputName='query'}, topNMetricSpec=NumericTopNMetricSpec{metric='PV'}, threshold=1000, querySegmentSpec=MultipleSpecificSegmentSpec{descriptors=[SegmentDescriptor{interval=2017-04-02T00:00:00.000+08:00/2017-04-02T23:59:59.999+08:00, version='2017-04-03T17:47:15.029+08:00', partitionNumber=48}, SegmentDescriptor{interval=2017-04-02T00:00:00.000+08:00/2017-04-02T23:59:59.999+08:00, version='2017-04-03T17:47:15.029+08:00', partitionNumber=157}, SegmentDescriptor{interval=2017-04-05T00:00:00.000+08:00/2017-04-05T23:59:59.999+08:00, version='2017-04-06T19:17:15.467+08:00', partitionNumber=61}, SegmentDescriptor{interval=2017-04-06T00:00:00.000+08:00/2017-04-06T23:59:59.999+08:00, version='2017-04-07T17:17:16.054+08:00', partitionNumber=163}, SegmentDescriptor{interval=2017-04-09T00:00:00.000+08:00/2017-04-09T23:59:59.999+08:00, version='2017-04-10T18:47:15.369+08:00', partitionNumber=120}, SegmentDescriptor{interval=2017-04-11T00:00:00.000+08:00/2017-04-11T23:59:59.999+08:00, version='2017-04-12T20:47:13.851+08:00', partitionNumber=98}, SegmentDescriptor{interval=2017-04-11T00:00:00.000+08:00/2017-04-11T23:59:59.999+08:00, version='2017-04-12T20:47:13.851+08:00', partitionNumber=120}, SegmentDescriptor{interval=2017-04-12T00:00:00.000+08:00/2017-04-12T23:59:59.999+08:00, version='2017-04-13T19:10:41.400+08:00', partitionNumber=142}, SegmentDescriptor{interval=2017-04-14T00:00:00.000+08:00/2017-04-14T23:59:59.999+08:00, version='2017-04-16T01:47:15.476+08:00', partitionNumber=52}, SegmentDescriptor{interval=2017-04-16T00:00:00.000+08:00/2017-04-16T23:59:59.999+08:00, version='2017-04-18T21:17:13.746+08:00', partitionNumber=224}, SegmentDescriptor{interval=2017-04-18T00:00:00.000+08:00/2017-04-18T23:59:59.999+08:00, version='2017-04-19T20:47:14.844+08:00', partitionNumber=72}]}, dimFilter=((stat_date = 20170331 || stat_date = 20170401 || stat_date = 20170402 || stat_date = 20170403 || stat_date = 20170404 || stat_date = 20170405 || stat_date = 20170406 || stat_date = 20170407 || stat_date = 20170408 || stat_date = 20170409 || stat_date = 20170410 || stat_date = 20170411 || stat_date = 20170412 || stat_date = 20170413 || stat_date = 20170414 || stat_date = 20170415 || stat_date = 20170416 || stat_date = 20170417 || stat_date = 20170418) && !query =  && cate_id = 350213), granularity='AllGranularity', aggregatorSpecs=[CardinalityAggregatorFactory{name='UV', fields='[DefaultDimensionSpec{dimension='visitor_id', outputName='visitor_id'}]'}, LongSumAggregatorFactory{fieldName='se_lpv_1d_004', name='PV'}, LongSumAggregatorFactory{fieldName='se_ipv_1d_003', name='IPV'}, DoubleSumAggregatorFactory{fieldName='pay_ord_amt_1d_007', name='成交金额'}, LongSumAggregatorFactory{fieldName='pay_ord_cnt_1d_007', name='成交笔数'}, DoubleSumAggregatorFactory{fieldName='pay_ord_amt_1d_019', name='全引导成交金额'}, LongSumAggregatorFactory{fieldName='pay_ord_cnt_1d_021', name='全引导成交笔数'}, LongSumAggregatorFactory{fieldName='clt_itm_cnt_1d_001', name='收藏商品数'}, LongSumAggregatorFactory{fieldName='cart_itm_cnt_1d_001', name='加购商品数'}, LongSumAggregatorFactory{fieldName='pay_ord_cnt_1d_024', name='C宝贝成交笔数'}, DoubleSumAggregatorFactory{fieldName='pay_ord_amt_1d_025', name='C宝贝成交金额'}, LongSumAggregatorFactory{fieldName='se_ipv_1d_005', name='C宝贝IPV'}, LongSumAggregatorFactory{fieldName='c_pv', name='C宝贝PV'}, LongSumAggregatorFactory{fieldName='b_pv', name='B宝贝PV'}, LongSumAggregatorFactory{fieldName='se_ipv_1d_004', name='B宝贝IPV'}], postAggregatorSpecs=[]}, peer=11.251.156.143}}
> {feed=alerts, timestamp=2017-04-20T14:16:03.894+08:00, service=historical, host=hadoop0797.et2.tbsite.net:8083, severity=component-failure, description=Exception handling request, data={class=io.druid.server.QueryResource, exceptionType=class java.nio.BufferUnderflowException, exceptionMessage=null, exceptionStackTrace=java.nio.BufferUnderflowException
>         at java.nio.Buffer.nextGetIndex(Buffer.java:506)
>         at java.nio.HeapByteBuffer.getShort(HeapByteBuffer.java:310)
>         at io.druid.query.aggregation.hyperloglog.HyperLogLogCollector.fold(HyperLogLogCollector.java:398)
>         at io.druid.query.aggregation.cardinality.CardinalityAggregatorFactory.combine(CardinalityAggregatorFactory.java:198)
>         at io.druid.query.topn.TopNBinaryFn.apply(TopNBinaryFn.java:106)
>         at io.druid.query.topn.TopNBinaryFn.apply(TopNBinaryFn.java:39)
>         at io.druid.common.guava.CombiningSequence$CombiningYieldingAccumulator.accumulate(CombiningSequence.java:212)
>         at com.metamx.common.guava.BaseSequence.makeYielder(BaseSequence.java:105)
>         at com.metamx.common.guava.BaseSequence.toYielder(BaseSequence.java:82)
>         at io.druid.common.guava.CombiningSequence.toYielder(CombiningSequence.java:78)
>         at com.metamx.common.guava.MappedSequence.toYielder(MappedSequence.java:46)
>         at io.druid.query.CPUTimeMetricQueryRunner$1.toYielder(CPUTimeMetricQueryRunner.java:93)
>         at com.metamx.common.guava.Sequences$1.toYielder(Sequences.java:98)
>         at io.druid.server.QueryResource.doPost(QueryResource.java:231)

**Comment by gianm** [comment]
Do you have any segments that were created by one of the buggy versions (what #3578 fixed)? If so, since the bug led to bad segments on disk, you have to reindex the bad segments or else you'll keep seeing this error even in 'good' versions of Druid.

FWIW we never released an 'official' release with the bug, but you might have got it if you built from master at some point or used the earlier 0.9.2 RCs.

**Comment by kaijianding** [comment]
I didn't build the segment with any HLL columns, all columns are string dimensions.
HLL is used only in query phase. @gianm

**Comment by jihoonson** [comment]
A similar issue is reported [here](https://groups.google.com/forum/#!searchin/druid-user/hll|sort:relevance/druid-user/zj9Q8RBh4Y4/2RJXRmj2DQAJ).

**Comment by gianm** [comment]
Ah, @kaijianding, you're right. It does look like something that is still a bug.

**Comment by gianm** [comment]
#3560 was caused by corruption in the HLL buffer caused by concurrent access from the indexing thread and a querying thread of a realtime task. In this case, there's no indexing. I wonder if the cause is the same, but some other kind of concurrent access. Nothing comes to mind straight away.

Did this trace happen on a historical? Did you notice anything else special that might help figure out what is going on? Are you running with any patches in your build of Druid that aren't in the official 0.9.2 release?

**Comment by kaijianding** [comment]
That's me... @jihoonson I found google user group is less attention payed, then I raised it here again.

**Comment by kaijianding** [comment]
This issue happens on historical, what I pasted in the main thread is from historical log.
I did run with patches, but nothing about TopN, cardinality or HLL. @gianm

**Comment by gianm** [comment]
Hmm, not really sure what might have caused that. Do you use the "cardinality" aggregator with any query types other than topN? If so have you noticed if they are ok or not?

**Comment by kaijianding** [comment]
I only use cardinality aggregator in TopN query, I can easily reproduce this issue with TopN query with multi-value dimension (but visitor_id mentioned in main thread is not multi-value).
The __schema is multi-value dimension
```
{
    "queryType": "topN",
    "dataSource": "opentsdb_60m",
    "dimension": "&&name",
    "threshold": 50,
    "metric": "count",
    "granularity": {
        "timeZone": "+08:00",
        "period": "P1D",
        "type": "period"
    },
    "aggregations": [
        {
            "type": "cardinality",
            "name": "count",
            "fieldNames": ["__schema"]
        }
    ],
    "intervals": [
        "2017-04-23/2017-04-24"
    ]
}
```
But I roughly tested similar query with groupBy query, this issue didn't happen. @gianm

**Comment by gianm** [comment]
What happens if you switch the dimension to some kind of extraction dimensionSpec? (it uses a different algorithm that treats the aggregators somewhat differently)

Are the stack traces always in `TopNBinaryFn.apply`?

**Comment by kaijianding** [comment]
Still error with extraction
```
{
    "queryType": "topN",
    "dataSource": "opentsdb_60m",
    "dimension": {
        "type": "extraction",
        "dimension": "&&name",
        "outputName": "&&name",
        "extractionFn": {
            "type": "identity"
        }
    },
    "threshold": 50,
    "metric": "count",
    "granularity": {
        "timeZone": "+08:00",
        "period": "P1D",
        "type": "period"
    },
    "aggregations": [
        {
            "type": "cardinality",
            "name": "count",
            "fieldNames": [
                "__schema"
            ]
        }
    ],
    "intervals": [
        "2017-04-23/2017-04-23T05"
    ]
}
```

Yes, always `TopNBinaryFn.apply`

**Comment by kaijianding** [comment]
new trace when extracting from cache @gianm 

> 2017-05-15T15:37:49,527 ERROR [background-cacher-4] io.druid.client.CachingQueryRunner - Error while getting future for cache task
> java.lang.RuntimeException: com.fasterxml.jackson.databind.JsonMappingException: (was java.nio.BufferOverflowException) (through reference chain: java.util.ArrayList[211]->java.util.ArrayList[1]->io.druid.query.aggregation.hyperloglog.HLLCV1["toByteArray()"])
>         at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[guava-16.0.1.jar:?]
>         at io.druid.client.CacheUtil.populate(CacheUtil.java:76) ~[druid-server-0.9.2.jar:0.9.2]
>         at io.druid.client.CachingQueryRunner$3.run(CachingQueryRunner.java:191) [druid-server-0.9.2.jar:0.9.2]
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1147) [?:1.8.0_102]
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622) [?:1.8.0_102]
>         at java.lang.Thread.run(Thread.java:834) [?:1.8.0_102]
> Caused by: com.fasterxml.jackson.databind.JsonMappingException: (was java.nio.BufferOverflowException) (through reference chain: java.util.ArrayList[211]->java.util.ArrayList[1]->io.druid.query.aggregation.hyperloglog.HLLCV1["toByteArray()"])
>         at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:210) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:177) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.std.JsonValueSerializer.serialize(JsonValueSerializer.java:184) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:100) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:21) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.std.AsArraySerializerBase.serialize(AsArraySerializerBase.java:183) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:100) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:21) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.std.AsArraySerializerBase.serialize(AsArraySerializerBase.java:183) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:128) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:2881) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsBytes(ObjectMapper.java:2362) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at io.druid.client.CacheUtil.populate(CacheUtil.java:61) ~[druid-server-0.9.2.jar:0.9.2]
>         ... 4 more
> Caused by: java.nio.BufferOverflowException
>         at java.nio.Buffer.nextPutIndex(Buffer.java:527) ~[?:1.8.0_102]
>         at java.nio.HeapByteBuffer.putShort(HeapByteBuffer.java:321) ~[?:1.8.0_102]
>         at io.druid.query.aggregation.hyperloglog.HyperLogLogCollector.toByteBuffer(HyperLogLogCollector.java:472) ~[druid-processing-0.9.2.jar:0.9.2]
>         at io.druid.query.aggregation.hyperloglog.HyperLogLogCollector.toByteArray(HyperLogLogCollector.java:486) ~[druid-processing-0.9.2.jar:0.9.2]
>         at sun.reflect.GeneratedMethodAccessor124.invoke(Unknown Source) ~[?:?]
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_102]
>         at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_102]
>         at com.fasterxml.jackson.databind.ser.std.JsonValueSerializer.serialize(JsonValueSerializer.java:155) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:100) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:21) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.std.AsArraySerializerBase.serialize(AsArraySerializerBase.java:183) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:100) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:21) ~[jackson-databind-2.4.6.jar:2.4.6]
>         at com.fasterxml.jackson.databind.ser.std.AsArraySerializerBase.serialize(AsArraySerializerBase.java:183) ~[jackson-databind-2.4.6.jar:2.4.6]

**Comment by gianm** [comment]
@kaijianding, do you know of a dataset/workload that could reproduce this on a single machine setup?

**Comment by gianm** [comment]
@kaijianding could you please check if the patch in #4304 helps? I'm not sure if it will, but it would protect against some un-synchronized uses of the HyperLogLogCollector returned by the cardinality aggregator.

Btw, I spoke to @drcrallen and he suggested one possible non-thread-safe use is background caching. Do you have that enabled?

**Comment by kaijianding** [comment]
Will try #4304 . I'm using background caching, @gianm It seems this issue is gone if I set populateCache and useCache to both false in query context

**Comment by gianm** [comment]
@kaijianding, if you're using background caching, you need #3943 too. That's in 0.10.0.

**Comment by kaijianding** [comment]
@gianm I know...... I raised PR #3943

**Comment by gianm** [comment]
Oh wait. That was your PR, so you probably know it already :)

Fwiw, there might be other bugs lurking in background caching -- the config isn't in the docs since we had seen some stability problems when enabling it. If you keep using it and find more that might be related, let us know…

**Comment by gianm** [comment]
Yeah, I just saw that, sorry :)

**Comment by gianm** [comment]
I think #4304 will probably help; it works around problems caused when reading a HyperLogLogCollector object from multiple threads at once. Background caching would do that, when it's being read into the cache at the same time as the main query thread is reading it.

**Comment by kaijianding** [comment]
I'm rolling upgrade the cluster now with #4304, will update you after several days run. I believe #4304 will help.
I'm using background caching with numBackgroundThreads=2, everything seems fine except for this issue. I think we can put this configuration in document, it's reliable, from my perspective.

**Comment by gianm** [comment]
Thanks @kaijianding. I want to check with @drcrallen first to see if he could describe the problems he noticed with background caching.

@drcrallen -- do you recall what you had seen with background caching? Do you think #3943 would have been enough to fix it, or were you seeing other kinds of problems?

**Comment by drcrallen** [comment]
https://github.com/druid-io/druid/pull/3943#issuecomment-280534185 is about the limit of what I remember without more thorough testing. Basically I couldn't get a good way to prove query resources were or were not released by the time the cache future ran.

**Comment by gianm** [comment]
Closed by #4304. @kaijianding, please let us know if you see the issue reoccur.
